cloudfront/api.wellcomecollection.org/send_slack_alert_for_5xx_errors.js

```diff
@@ -0,0 +1,339 @@
+/** Identify errors in the CloudFront logs and post them to Slack.
+ *
+ * Our CloudFront logs get written to an S3 bucket.  This is the source code
+ * of a Lambda function that:
+ *
+ *    1. Listens to the event stream from the bucket
+ *    2. Gets notified of new logs being written to the bucket
+ *    3. Reads new logs, and filters for 5xx errors
+ *    4. If there are any, sends a list of those URLs to a shared Slack channel
+ *
+ * This tells us if something is broken, and helps us target our debugging.
+ *
+ * (As opposed to our previous approach, which dropped a vague metric telling us
+ * there were errors, but with no way to identify which app was the source of
+ * the errors.  Now we know which app's logs we should be checking first.)
+ *
+ * It's written in vanilla JavaScript/Node because it's pretty simple, and
+ * it simplifies the process of deployment.
+ *
+ */
+
+const AWS = require('aws-sdk');
+const events = require('events');
+const https = require('https');
+const readline = require('readline');
+const zlib = require('zlib');
+
+/** Reads a CloudFront log file and returns structured log events.
+ *
+ * CloudFront logs are semi-structured files, where the top two lines are
+ * a header, then subsequent lines are tab-separated values, e.g.:
+ *
+ *      #Version: 1.0
+ *      #Fields: date time cs-uri-stem
+ *      2022-05-18	14:11:10	/works
+ *      2022-05-18	14:11:09	/works/atry66dj/items
+ *
+ * This function emits objects so that downstream callers don't need
+ * to know about the structure of this file, e.g.
+ *
+ *      { date: '2022-05-18', time: '14:11:10', cs-uri-stem: '/search/works' }
+ *      { date: '2022-05-18', time: '14:11:09', cs-uri-stem: '/works/atry66dj/items' }
+ *
+ */
+async function findCloudFrontHitsFromLog(bucket, key) {
```

Comment:
The implementation of this is strongly Python-developery 🐍

Reply:
Well it was originally written in Snekk-script. 😜 

The only reason it's in JavaScript now is that I was putting it in the front-end repo where there's very little (any?) Python and it felt like introducing another language was creating an unnecessary complexity.